### PR TITLE
OCI Native Generative AI Example

### PR DESCRIPTION
**Pull request by bhoran**
An example of using the OCI generative AI service, using the native image maven plugin

### CODE REVIEW

File: native-oci-generative-ai/README.md
Comment (olyagpl) [inline native-oci-generative-ai/README.md:1]: Once the PR is ready for merge, please add an entry for this demo in the top-level README.

Comment (fniephaus) [inline native-oci-generative-ai/README.md:1]: Please add a GitHub workflow so that this demo is tested.

File: native-oci-generative-ai/README.md
Comment (fniephaus) [inline native-oci-generative-ai/README.md:5]:
```suggestion
The [source code](./src/main/java/com/oracle/labs/GenerateTextExample.java) is a self-contained demonstration of using a [`GenerativeAiInferenceClient`](https://docs.oracle.com/iaas/tools/java/latest/com/oracle/bmc/generativeaiinference/GenerativeAiInferenceClient.html) to generate text, based on a prompt.
```

File: native-oci-generative-ai/README.md
Comment (fniephaus) [inline native-oci-generative-ai/README.md:25]: These could just be Java properties or environment variables, which can then be set at run-time (or frozen into the executable via build-time init if you insist)?

File: native-oci-generative-ai/pom.xml
Comment (fniephaus) [inline native-oci-generative-ai/pom.xml:9]: Should this include the word "Demo" somewhere?
```suggestion
  <name>Oracle Cloud Infrastructure - Generative AI Demo</name>
```

File: native-oci-generative-ai/pom.xml
Comment (fniephaus) [inline native-oci-generative-ai/pom.xml:19]: One too many new lines:
```suggestion
```

File: native-oci-generative-ai/pom.xml
Comment (fniephaus) [inline native-oci-generative-ai/pom.xml:81]: This demo does not seem to be using a test phase, so this `<execution>` entry could be left out if you want to keep the pom.xml as concise as possible.

File: native-oci-generative-ai/pom.xml
Comment (fniephaus) [inline native-oci-generative-ai/pom.xml:109]: Is lombok really necessary?

File: native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java
Comment (fniephaus) [inline native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java:96]: The line break is a bit odd here. Maybe use `/*  */` for a multiline comment?

File: native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java
Comment (fniephaus) [inline native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java:117]: Use a text block (and, unlike me, an IDE for formatting :wink:)
```suggestion
        .prompt("""
            Generate a job description for a data visualization expert with the following three qualifications only:
            1) At least 5 years of data visualization expert
            2) A great eye for details
            3) Ability to create original visualizations")
        """)
```

File: native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java
Comment (fniephaus) [inline native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java:121]: Use double literals?
```suggestion
        .temperature(0.5d)
        .frequencyPenalty(1.0d)
        .topP(0.75d)
```

File: native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java
Comment (fniephaus) [inline native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java:114]: Maybe split the comment to highlight where the request building stop, and the sending happens?

File: native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java
Comment (fniephaus) [inline native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java:137]: I'm guessing this is the send and get, right?

File: native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java
Comment (fniephaus) [inline native-oci-generative-ai/src/main/java/com/oracle/labs/GenerateTextExample.java:138]: This should just do the right thing, even if `generateTextResponse` can ever be `null` (not sure it can):
```suggestion
    System.out.println(generateTextResponse);
```

**Review by fniephaus — CHANGES_REQUESTED** [review]
Thanks for the demo. LGTM so far, left a couple of minor comments. I'll approve when there's a workflow that tests the demo :wink:

File: native-oci-generative-ai/README.md
Comment (fniephaus) [inline native-oci-generative-ai/README.md:55]: Could you maybe add sample output so that users see what's supposed to happen when running the demo (either on HotSpot or in native mode).

File: native-oci-generative-ai/README.md
Comment (fniephaus) [reply native-oci-generative-ai/README.md:1]: Right. We could do this via [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions), but in this case, I would be ok with just building the app with Native Image.

File: native-oci-generative-ai/README.md
Comment (olyagpl) [inline native-oci-generative-ai/README.md:13]: no empty line is needed between *

File: native-oci-generative-ai/README.md
Comment (bhoran) [reply native-oci-generative-ai/README.md:87]: Done

File: native-oci-generative-ai/pom.xml
Comment (olyagpl) [inline native-oci-generative-ai/pom.xml:14]: Here Java version is 21, while in the workflow it's 17. They should match, because the action is failing: 
`Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project generative-ai: Fatal error compiling: error: invalid target release: 21 -> [Help 1] `

**Review by olyagpl — APPROVED** [review]

**Comment by bhoran** [comment]
@fniephaus -- I've created the workflow and it passes.
I haven't updated the source as you suggested because I want it to closely match the source that is produced by https://docs.oracle.com/iaas/Content/generative-ai/use-playground-generate.htm.

File: .github/workflows/native-oci-generative-ai.yml
Comment (fniephaus) [inline .github/workflows/native-oci-generative-ai.yml:25]: Can we also test this against `latest-ea` to ensure updates don't break the demo? Also, the latest JDK 21 release is `21.0.3`.

File: native-oci-generative-ai/pom.xml
Comment (fniephaus) [inline native-oci-generative-ai/pom.xml:15]: Is there a technical reason to lock the Java version? I think we can just leave it out so it's using the Java version of the `JAVA_HOME` provided:
```suggestion
```

File: native-oci-generative-ai/pom.xml
Comment (fniephaus) [inline native-oci-generative-ai/pom.xml:29]: Same here, do we need to lock the Java version?
```suggestion
        <configuration>
          <source>21</source>
          <target>21</target>
        </configuration>
```

**Review by fniephaus — APPROVED** [review]
LGTM

**Comment by fniephaus** [comment]
> I've created the workflow and it passes.
I haven't updated the source as you suggested because I want it to closely match the source that is produced by https://docs.oracle.com/iaas/Content/generative-ai/use-playground-generate.htm.

@bhoran great, thanks! and yes, that makes sense.

File: .github/workflows/native-oci-generative-ai.yml
Comment (fniephaus) [inline .github/workflows/native-oci-generative-ai.yml:31]: Use `-B`/`--batch-mode` or at least `--no-transfer-progress` for shorter build logs ([problematic example](https://github.com/graalvm/graalvm-demos/actions/runs/9061198825/job/24892496069?pr=264#step:4:76)):
```suggestion
          mvn --batch-mode clean package -Pnative
```

**Review by olyagpl — APPROVED** [review]

**Review by fniephaus — APPROVED** [review]
LGTM

**Pull request merged**
